plot-lib/plot/private/plot2d/interval.rkt

```diff
@@ -374,3 +375,89 @@
                                                line1-color line1-width line1-style
                                                line2-color line2-width line2-style
                                                alpha))]))
+
+;; ===================================================================================================
+;; Violin
+
+(:: violin
+    (->* [Real (Sequenceof Real)
+          (U Real #f) (U Real #f)]
+         [#:x-min (U Real #f) #:x-max (U Real #f)
+          #:samples Positive-Integer
+          #:color Plot-Color
+          #:style Plot-Brush-Style
+          #:line1-color Plot-Color
+          #:line1-width Nonnegative-Real
+          #:line1-style Plot-Pen-Style
+          #:line2-color Plot-Color
+          #:line2-width Nonnegative-Real
+          #:line2-style Plot-Pen-Style
+          #:alpha Nonnegative-Real
+          #:label (U String pict #f)
+          #:bandwidth (U Real #f)
+          #:invert? Boolean]
+         renderer2d))
+(define (violin
+         x-center ys
+         [y-min #f] [y-max #f]
+         #:x-min [x-min #f] #:x-max [x-max #f]
```

Comment:
I am not sure what the purpose of the `x-min` and `x-max` arguments are.  For a renderer they should control the X range, if it makes sense.   Also, there seems not be a way to control the width of the violin plot (seems to be a width of 1 from -0.5 to 0.5?)

Since the violin plots have essentially one dimension, we need a mechanism to allow the user to reliably place several of them next to each other, similar to how `discrete-histogram` uses  `discrete-histogram-skip` and `discrete-histogram-gap`.  So basically:

* add `#:skip` and `#:gap` parameters to the `violin` plot, they can default to the same `discrete-histogram-skip` and `discrete-histogram-gap`
* the width of the plot should be (- skip gap), so the function interval should be scaled by this amount

Reply:
In general violin plots can have arbitrary width because the KDE can have arbitrary width (the constraint is that the area is 1 but nothing on the maximum width). Taking a look at other plotting libraries it looks like some have an option to do either a fixed width (by scaling) or the full width. So maybe we have a `#:width` that defaults to `(- skip gap)` and if it's explicitly `#f` to use the full width.

As for several violins, do you think it would make more sense to have the data be more in the format of `stacked-histogram`? I.e. instead of just ploting one violin you would do: `(violin (list #("a" (1 2 3)) #("b" (3 4 5)) ...)` and you would get several next to one another. Then `#:skip` and `#:gap` make more sense.

Reply:
The value `(- skip gap)` should always represent the width of 1.0 for a violin plot and the plot might extend that width, possibly overlapping with nearby plots.  The requested bounds for the renderer should always be the actual width of the plot (scaled by `(- skip gap)`), so the plot is auto-sized to fit the whole plot by default, with no clipping.  If the user does not want their violin plots to overlap, they can specify larger `x` values between adjacent `violin` plots.  Is there a case where an extra `width` argument is required?

The `skip` and `gap` arguments are most useful when the `violin` plot is used with other renderers and the plot area is not "unit" in size. 

> As for several violins, do you think it would make more sense to have the data be more in the format of stacked-histogram? 

I think it is simpler to just have `violin` plot a single data set, as it does now:

* it is easy to add multiple violin plots and also to use `for/list` to automatically create a multi-violin plot.  
* most novice users will have trouble specifying the data in a correct format, since it would involve using nested lists.  
* for presentation purposes, it is common to use different colors for each violin in a multi-violin plot, so the input data would be even more complex, or users are back to using multiple `violin` renderers.

I think we should keep `violin` as it is now, plotting a single data set.  In the future, we can add a new `violin*` renderer if users request it.